### PR TITLE
Allow uri's in the config file

### LINKED ISSUE #537: Allow the config file to have a uri instead of host-port
Host port doesn't allow us to configure the authentication or load balancing that is required for connection to the central mongo cluster. Currently databroker only support host-port.

https://api.mongodb.com/python/current/examples/authentication.html#authentication-examples

### PR DESCRIPTION
**Pull request by gwbischof**
Fixes #537

### CODE REVIEW

**Comment by gwbischof** [comment]
Should of made this a draft. It is still a work in progress.

**Comment by danielballan** [comment]
Looks on the right track, once you can satisfy the crufty v1 stuff that is failing tests. :-\

**Comment by gwbischof** [comment]
Still need to figure out how to make this work for V0 databroker

**Comment by gwbischof** [comment]
I think it is ready for review now

File: databroker/assets/mongo.py [hunk 1]
```diff
@@ -104,10 +102,9 @@ def _datum_col(self):
 
     @property
     def _connection(self):
-        if self.__conn is None:
-            self.__conn = MongoClient(self.config['host'],
-                                      self.config.get('port', None))
-        return self.__conn
+        if self.__db is None:
+            self.__db = _get_mongo_database(self.config)
+        return self.__db.client
```
Comment (danielballan) [inline databroker/assets/mongo.py]: Can we use `self._db.client` and skip the `None` check above?

File: databroker/headersource/mongo.py [hunk 2]
```diff
@@ -49,24 +45,14 @@ def reconfigure(self, config):
...
-                    self.config['host'],
-                    self.config['port'])
-                self.__conn = MongoClient(uri)
-            else:
-                self.__conn = MongoClient(self.config['host'],
-                                          self.config.get('port', None))
-        return self.__conn
+        if self.__db is None:
+            self.__db = _get_mongo_database(self.config)
+        return self.__db.client
```
Comment (danielballan) [inline databroker/headersource/mongo.py]: Same here

File: databroker/assets/mongo.py (hunk 1 above)
Comment (gwbischof) [reply databroker/assets/mongo.py]: I don't understand the reason for this change, can you explain?

Comment (danielballan) [reply databroker/assets/mongo.py]: If we access `self._db`, it perhaps the `if self.__db is None: ...` check  itself, and we can reduce the repeated code.

Comment (gwbischof) [reply databroker/assets/mongo.py]: updated

File: databroker/headersource/mongo.py (hunk 2 above)
Comment (gwbischof) [reply databroker/headersource/mongo.py]: updated

**Pull request merged**